includes/class-wc-cart-session.php

```diff
@@ -128,7 +128,16 @@ public function get_cart_from_session() {
 				$update_cart_session = true;
 				do_action( 'woocommerce_remove_cart_item_from_session', $key, $values );
 
-			} elseif ( ! $product->is_purchasable() ) {
+			/**
+			 * Allow 3rd parties to override this item's is_puchasable() result with cart item data.
+			 *
+			 * @since 4.1.0
+			 *
+			 * @param bool $is_purchasable If false, the item will not be added to the cart. Default: product's is_purchasable() status.
+			 * @param string $key Cart item key.
+			 * @param array $values Cart item values e.g. quantity and product_id.
+			 */
+			} elseif ( ! apply_filters( 'woocommerce_cart_item_is_purchasable', $product->is_purchasable(), $key, $values ) ) {
```

Comment:
You can achieve it already with `woocommerce_pre_remove_cart_item_from_session`.
If you like we could change to `apply_filters( 'woocommerce_pre_remove_cart_item_from_session', false, $key, $values, $product )` so you don't need to load the product data again.

Reply:
Hi @claudiosanches that might work...

Reply:
Though if we're bypassing the `is_purchasable()` check and overriding it's value to keep the product in the cart then we don't need to set `$update_cart_session = true;` if the product isn't getting removed.

Reply:
In both places will set `$update_cart_session`, but with the pre filter you can include any notification you like.

Reply:
If the product is not removed though, why would the cart session need to update? I'm trying to bypass the removal and leave the product in the cart (if it's part of a bundle). Though I do need to think if there's another way to do this.

Reply:
> If the product is not removed though, why would the cart session need to update?

By default it's there, but if it's a problem, just fine to customize it.
`woocommerce_pre_remove_cart_item_from_session` is the way to customize this method, if necessary we can improve it.

Reply:
@helgatheviking do you think if we pass `$product` in the `woocommerce_pre_remove_cart_item_from_session` filter will work for you?